### PR TITLE
Make the trace compiler comprehendable

### PR DESCRIPTION
**Pull request by vext01**
The first commit is what I set out to do.

The others are things I noticed along the way that I thought should be separate.

See commit messages.

(this will be hard to review, sorry!)

### CODE REVIEW

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -49,6 +50,235 @@ uint64_t getNewTraceIdx() {
...
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  // Return the instruction immediately before the instruction to resume at.
+  //
+  // Throws if there is no resume point set.
+  Instruction *getResumeAfterInstruction() {
+    return (*Resume).ResumeAfterInstr;
```
Comment (vext01) [inline ykllvmwrap/src/jitmodbuilder.cc:179]: Oops. This should use `->value()` instead of `*`  to ensure it throws.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 2]
```diff
@@ -49,6 +50,235 @@ uint64_t getNewTraceIdx() {
 // The name prefix used for blocks that are branched to when a guard succeeds.
 #define GUARD_SUCCESS_BLOCK_NAME "guardsuccess"
 
+// Describes a "resume point" in an LLVM IR basic block.
+//
+// This is required because in LLVM IR, a call instruction does not terminate a
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:55]: Can we make this a bit easier to parse? Maybe something like "Since call instructions are not terminators in LLVM IR, we have to keep track of which call within a block we last processed."

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 3]
```diff
@@ -49,6 +50,235 @@ uint64_t getNewTraceIdx() {
...
+  // If we want to resume at the `load`, then we store the `call`.
+  Instruction *ResumeAfterInstr;
+  // The index of the instruction immediately *before* where we should resume.
+  //
+  // This could be computed by looping over `ResumeInstr`'s parent block, but
+  // since it's needed fairly frequently, we cache it.
+  size_t ResumeAfterInstrIdx;
+
+  // Dump the resume point to stderr for debugging.
+  void dump() {
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:80]: Can we remove this debugging code?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 4]
```diff
@@ -49,6 +50,235 @@ uint64_t getNewTraceIdx() {
...
+  // progress through the frame's function (or null if no stackmap call has
+  // been encountered in this frame yet).
+  CallInst *LastSMCall;
+  // The resume point for this frame (if any).
+  std::optional<BlockResumePoint> Resume;
+  // When true, this frame initiated outlining. When this frame is the
+  // most-recent frame again, outlining will cease.
+  bool OutlineBase = false;
+
+  // Dump the frame to stderr for debugging.
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:108]: Ditto.

**Comment by ltratt** [comment]
I'm not keen on all these `dump` methods: they feel like left-over debugging code; they're not conditionally compiled in; and it would be better if they returned strings, rather than dumping stuff to stderr. I can maybe be convinced that they shouldn't be nuked, but I think an explicit argument needs to be made for them.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 5]
```diff
@@ -1275,46 +1511,70 @@ class JITModBuilder {
...
     for (size_t Idx = 0; Idx < InpTrace.Length(); Idx++) {
+      // Update the previously executed BB in the most-recent frame (if it's
+      // mappable).
       optional<IRBlock> MaybeIB = InpTrace[Idx];
 
       // If we previously saw a call to foreign code, there will now be an
       // "unmappable hole" in the trace.
       if (ExpectUnmappable) {
         assert(!MaybeIB.has_value());
+        // There are never two unmappable regions in a row, and since we don't
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:1477]: Sentence is hard to parse.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 6]
```diff
@@ -1564,17 +1564,20 @@ class JITModBuilder {
...
-      for (size_t CurInstrIdx = 0; CurInstrIdx < BB->size(); CurInstrIdx++) {
-        // If we've returned from a call, skip ahead to the instruction where
-        // we left off.
-        std::optional<size_t> ResumeIdx = MPF->getResumeInstrIdx();
-        if (ResumeIdx.has_value()) {
-          CurInstrIdx = *ResumeIdx;
-          MPF->clearResume();
-        }
-
+      for (; CurInstrIdx < BB->size(); CurInstrIdx++) {
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:1533]: I'm generally not a fan of non-standard `for`s: I think this would be less likely to be misread as a `while`.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:80]: Why? You wouldn't object to a `impl Debug` implementation in Rust. This is equivalent. It's all nicely encapsulated and we even follow the LLVM convention of naming the functions "dump".

(I grow tired of repeatedly writing (non-trivial) debugging routines every time I want to do some debugging. C++ isn't like rust, where you can easily derive `Debug` printing -- everything has to be done manually, and it's very laborious)

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 6 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:1533]: If we make it a `while` then each `continue` will need an accompanying `CurInsrIdx++` which I think is even more risky than the non-standard while.

I count 7 `continue`s in the while loop.

**Comment by vext01** [comment]
> it would be better if they returned strings

The LLVM precedent is to have a [`print()`](https://llvm.org/doxygen/classllvm_1_1Value.html#a1e56c814d11206720cc23059b871128d) method which writes the output to a stream (which could be a string, or a file), and a [`dump()`](https://llvm.org/doxygen/classllvm_1_1Value.html#af7dca9a9e816ef69fd9e9467f64f72b4) which uses the aforementioned to write to the `llvm::errs()` stream (a.k.a. stderr).

We always just use this stuff for print debugging, so I skipped `print()` .

**Comment by vext01** [comment]
> they're not conditionally compiled in

Why would they be? You don't conditionally compile in `impl Debug` in Rust :)

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:55]: I think that throws away too much information.

Asides from the typo, which part is hard to parse, and I'll see what I can do.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 1 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc]: Fixed in 3bea2eea638eadadc3a0207825258223e0dfcb16

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc]: Fixed in f44ae80727afd814e0c221542e9aa2ddccb838fd

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 6 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:1538]: Fair enough!

**Comment by ltratt** [comment]
Only one of my comments remains open.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 7]
```diff
@@ -49,6 +50,235 @@ uint64_t getNewTraceIdx() {
...
+  //
+  // To be clear, if we have two instructions:
+  // ```
+  //   call @f()
+  //   %x = load(...)
+  // ```
+  //
+  // If we want to resume at the `load`, then we store the `call`.
+  Instruction *ResumeAfterInstr;
+  // The index of the instruction immediately *before* where we should resume.
```
[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: To be clear, it is correct that this refers to the index of the instruction in the parent basic block.

(Nothing is byte-centric when working with the LLVM API)

[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: The index of its byte or ... ?

I was actually referring more to the "*before*" bit -- I think that's wrong?

[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: If we have a BB like:
```
bb0:
  %1 = alloca ...
  %2 = add ...
  %3 = call @f(...)
  %4 = sub ...
  ret
```

And we want to talk about the call, that's instruction index 2 of `bb0`, and we will resume at index 3.

In some places we add one to talk about the instruction to resume at, in others we don't to talk about the call itself.

So "before" is correct.

[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: > In some places we add one to talk about the instruction to resume at, in others we don't to talk about the call itself.
> 
> So "before" is correct.

But what you've said there is that sometimes we refer to `%3` (which I agree is "before") but sometimes `%4` (which is "after").

[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: Yes, that's right.

Maybe this accessor makes it clearer?

[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: The accessor suggests it would be either `%4` or `null`?

[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]:
```
// The index of the instruction immediately *before* where we should resume.
size_t ResumeAfterInstrIdx;
```

```
bb0:
  %1 = alloca ...
  %2 = add ...
  %3 = call @f(...)
  %4 = sub ...
  ret
```

We should resume after the call at index 3, so `ResumeAfterInstrIndex` would be 2. The accessor would add one to get index 3.

Is the confusion that this is a "resume after" data structure, not a "resume at" data structure?

[the remark this answers is not in the document]
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: To clarify: when we set the `ResumeAfterInstrIdx`, we are currently processing the `call` instr. When processing the call we now expect to jump to a different block, process that block, before then returning back to the current block. Upon returning, we want to resume after the last processed instruction, which is the instruction after the `call`.

[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: > I guess there is no reason why this couldn't be ResumeAtInstrIdx, but that would mean we just move the +1 to a different place (i.e. when we set this variable).

The problem with that is that one of the inner fields would point to the call, and another to the instruction after the call.

You could point them both at the instruction after the call, but then you'd need a loop to find the previous instruction.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 8]
```diff
@@ -273,24 +476,33 @@ class JITModBuilder {
...
+    MappableFrame *MPF = CallStack.curMappableFrame();
+    assert(MPF);
+    MPF->OutlineBase = true;
+    Outlining = true;
   }
 
-  // Determines whether we are currently outlining code, i.e. we are processing
-  // the IR of a function we do not wish to inline.
-  bool isOutlining() { return OutlineCallDepth > 0; }
+  // Check if we should stop outlining.
```
Comment (ptersilie) [inline ykllvmwrap/src/jitmodbuilder.cc:449]: From the description I assumed this would return a bool to tell us if we can stop outlining. I think the comment should be something like "Try to stop outlining if possible". And maybe we can also change the function name to `tryStopOutlining` to make it more obvious this isn't just a check, but changes the state.

**Comment by ptersilie** [comment]
Added some comments/typos.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: So we currently have:

> // The index of the instruction immediately *before* where we should resume.

I don't know how to be more explicit.

Does this help?:

> The index of the instruction prior to the resume point.

[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: Right, I get it now. But to avoid this, can we just say "// The index of the call instruction"?

[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: They may both be call instructions.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 9]
```diff
@@ -474,23 +689,23 @@ class JITModBuilder {
...
-    ActiveFrames.pop_back();
-    CallDepth -= 1;
-    ResumeAfter = InlinedCalls.back();
-    InlinedCalls.pop_back();
-    LastCompletedBlocks.pop_back();
-    if (isOutlining()) {
-      OutlineCallDepth -= 1;
+    CallStack.popFrame();
+
+    // Check if we have arrived back the frame where outlining started.
```
[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc]: Fixed in a431811831903a669204f2f985cc14a1cc1b1594

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 10]
```diff
@@ -632,25 +847,33 @@ class JITModBuilder {
...
     // FIXME: Use function index instead of string name.
-    ActiveFrames.push_back(FInfo); // Add current frame.
     StructType *ActiveFrameSTy = StructType::get(
         Context, {PointerSizedIntTy, PointerSizedIntTy, Int8PtrTy});
     AllocaInst *ActiveFrameVec = FailBuilder.CreateAlloca(
-        ActiveFrameSTy,
-        ConstantInt::get(PointerSizedIntTy, ActiveFrames.size()));
+        ActiveFrameSTy, ConstantInt::get(PointerSizedIntTy, CallStack.size()));
+
+    // To describe the callstack at a guard failure, The following code uses
```
[the remark this answers is not in the document]
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc]: Fixed in 040c1d14ff7c1f3f8fc5ae7f48611d15c4420d84

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: I'm still confused but I'm going to stop and mark this resolved because I'm not sure we're getting anywhere useful.

[the remark this answers is not in the document]
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:75]: It seems obvious to me that "the call" is the call being examined, so it doesn't seem ambiguous. But I want to move on.

**Comment by ptersilie** [comment]
I'm happy. If @ltratt has the same emotional response then I think this is ready for squashing.

**Comment by ltratt** [comment]
I have no emotions other than the joy of saying: please squash.

**Comment by vext01** [comment]
splat.

**Comment by ptersilie** [comment]
Let's goooo :car: 

bors r+
